### DISCUSSION TITLE
Regarding Usage of EWC strategy and Replay strategy

### DISCUSSION THREAD

**Discussion by AdityaJainTD** [ContinualAI/avalanche#1065] [Q&A] [answered]
Hi,

I had some doubts with respect to the usage of few of the functionalities. They are as follows:-
 
1. EWC strategy
-  I wanted to experiment with EWC strategy. The scenario I am using is nc_benchmark and my pseudocode is as shown below:-

    cl_strategy = EWC(model, optimizer, criterion ,ewc_lambda = 0.001, train_mb_size=1, train_epochs=2 , mode='separate', 
     eval_mb_size=1, device=device)
 
    for exp in train_stream:
                cl_strategy.train(exp)
 
   The evaluation accuracy is quite bad and close to the Naïve baseline method. The documentation says ''_Importance are computed 
   with an additional pass on the training set_". So I wanted to ask is the weight importance computed automatically or do I need to 
   add them explicitly in my training loop?

2. Replay strategy
- Similarly just like EWC strategy with nc_benchmark setting, I am also experimenting with the Replay strategy. I have kept the bin size 
   to the default setting of 200 . Lets say my target distribution is as follows:-

  Class | Samples
       -- | --
        1 | 1200
        2 | 900
        3 | 500
        4 | 300

  My assumption with replay strategy was that, it would maintain random samples of previous classes in an external memory. So lets 
  say when training on class 2, ideally the sample size should be 900 + 200 (sample of class 1), when training on class 3, the sample 
  size would be 500 + 200 (sample of 1 and 2). But I see only the exact samples in the training loop i.e. when training on class 2, the 
  sample size is 900 and so on. Can you please help me understand how this strategy works? Below is the pseudocode for the same:-
 
  cl_strategy = Replay( model, optimizer, criterion , train_mb_size=1, train_epochs=2, eval_mb_size=1, device=device) 
 
  for exp in train_stream:
                cl_strategy.train(exp)

**Comment by AndreaCossu** [comment, marked as answer]
1. EWC automatically computes importances. Importance-based regularization strategies are well-known to not work in New Classes scenario so your results are correct.
2. Replay keeps a fixed memory (of 200 previous samples in your cases) and balances it across the encountered classes. So each time, you will have in your memory `200/n_classes_encountered_so_far per` examples per class in your memory. Replay concatenates to the current minibatch a random sample from the memory and trains the model on this extended minibatch. 

Hope this helps :smile:

**Reply by AdityaJainTD** [reply]
Thankyou @AndreaCossu for the answer. I am still not 100% clear with the second answer. So, if the replay strategy concatenates a sample sample from the memory with the current mini-batch, ideally when I print 'exp.classes_in_this_experience' (exp being the sample from the training stream), I should see multiple classes as the output, right?

**Reply by AndreaCossu** [reply]
No, that regards the benchmark alone. The benchmark and the strategy are separate objects. `classes_in_this_experience ` represents a property of the benchmark. Replay modifies the dataset by including previous classes but the classes present in that experience are still the original ones. The others come from the replay buffer.

**Reply by AntonioCarta** [reply]
> Replay modifies the dataset by including previous classes but the classes present in that experience are still the original ones.

A small clarification about this: the original experience's dataset is not modified. The strategy uses a dataloader which combines data from a separate buffer and from the experience data.

**Reply by AdityaJainTD** [reply]
Okay, so the strategy internally takes care of training it on the random collected sample containing previous examples?

**Reply by AndreaCossu** [reply]
yes, Replay does that under the hood. It trains the model on a mix of examples from the memory and examples from the current minibatch.

**Reply by AdityaJainTD** [reply]
Thankyou so much for the response!
